### ISSUE TITLE
No functional channel service provider found. Try adding a dependency on the grpc-okhttp, grpc-netty, or grpc-netty-shaded artifact

### ISSUE THREAD

**Issue by cxMax** [grpc/grpc-java#9152]
## description
> Caused by io.grpc.ManagedChannelProvider$ProviderNotFoundException
No functional channel service provider found. Try adding a dependency on the grpc-okhttp, grpc-netty, or grpc-netty-shaded artifact

when I use firebase-firestore, there is an exception was occured.

## dependencies.
implementation 'com.google.firebase:firebase-bom:29.2.1'
implementation 'com.google.firebase:firebase-analytics:19.0.0'
implementation 'com.google.firebase:firebase-crashlytics:18.0.1'
implementation 'com.google.firebase:firebase-firestore:24.0.1'

## java stack:
Fatal Exception: java.lang.RuntimeException
Internal error in Cloud Firestore (24.0.1).
com.google.firebase.firestore.util.AsyncQueue.lambda$panic$3 (AsyncQueue.java:539)
com.google.firebase.firestore.util.-$$Lambda$AsyncQueue$cG4e8FKiBI1uHuEBYC-bY9JR88Y.run (-.java:2)
android.os.Handler.handleCallback (Handler.java:938)
android.os.Handler.dispatchMessage (Handler.java:99)
android.os.Looper.loop (Looper.java:236)
android.app.ActivityThread.main (ActivityThread.java:8073)
java.lang.reflect.Method.invoke (Method.java)
com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:620)
com.android.internal.os.ZygoteInit.main (ZygoteInit.java:1011)

Caused by com.google.android.gms.tasks.RuntimeExecutionException
io.grpc.ManagedChannelProvider$ProviderNotFoundException: No functional channel service provider found. Try adding a dependency on the grpc-okhttp, grpc-netty, or grpc-netty-shaded artifact
com.google.android.gms.tasks.zzu.getResult (zzu.java:15)
com.google.firebase.firestore.remote.FirestoreChannel.lambda$runBidiStreamingRpc$0 (FirestoreChannel.java:115)
com.google.firebase.firestore.remote.-$$Lambda$FirestoreChannel$0Ae0OZVH5uV1moEedILJLfrbyWY.onComplete (-.java:6)
com.google.android.gms.tasks.zzj.run (zzj.java:4)
java.util.concurrent.Executors$RunnableAdapter.call (Executors.java:462)
java.util.concurrent.FutureTask.run (FutureTask.java:266)
java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run (ScheduledThreadPoolExecutor.java:301)
java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1167)
java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:641)
com.google.firebase.firestore.util.AsyncQueue$SynchronizedShutdownAwareExecutor$DelayedStartFactory.run (AsyncQueue.java:234)
java.lang.Thread.run (Thread.java:923)

Caused by io.grpc.ManagedChannelProvider$ProviderNotFoundException
No functional channel service provider found. Try adding a dependency on the grpc-okhttp, grpc-netty, or grpc-netty-shaded artifact
io.grpc.ManagedChannelProvider.provider (ManagedChannelProvider.java:43)
io.grpc.ManagedChannelBuilder.forTarget (ManagedChannelBuilder.java:76)
com.google.firebase.firestore.remote.GrpcCallProvider.initChannel (GrpcCallProvider.java:113)
com.google.firebase.firestore.remote.GrpcCallProvider.lambda$initChannelTask$6 (GrpcCallProvider.java:256)
com.google.firebase.firestore.remote.-$$Lambda$GrpcCallProvider$-4U_AcdJVCEeG6NOrpg3tM4Nfcc.call (-.java:2)
com.google.android.gms.tasks.zzv.run (zzv.java:2)
com.google.firebase.firestore.util.ThrottledForwardingExecutor.lambda$execute$0 (ThrottledForwardingExecutor.java:54)
com.google.firebase.firestore.util.-$$Lambda$ThrottledForwardingExecutor$LidNF-W_9-SVM15Diteen_5-PO8.run (-.java:4)
java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1167)
java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:641)
java.lang.Thread.run (Thread.java:923)

## How did it happen
I sampely use firestore as the tutorial guide say， below is my usage

FirebaseFirestore.getInstance()
.collection("feedback")
.document("RLISPq6dD1H53XRUsx0k")
.collection(SimpleDateFormat("yyyy-MM-dd HH:mm:ss").format(System.currentTimeMillis()) + "_" + UUID.randomUUID())
.add(mapOf("extra" to getExtraParams()))

## I track the code `FirebaseFirestore.getInstance().collection()`, then I found how this crash happend
1. First Time, `FirebaseFirestore.ensureClientConfigured()` will create a new FirestoreClient Object
<img width="826" alt="image" src="https://user-images.githubusercontent.com/17964075/167250414-5a182ff7-408b-462c-a51e-7ac645ec6370.png">
2. `FirestoreClient.initialize(Context context, User user, FirebaseFirestoreSettings settings)` will create a new Datastore object
<img width="888" alt="image" src="https://user-images.githubusercontent.com/17964075/167250463-9085bfbf-0edf-445d-98f9-6f82e77ccf1d.png">
3. create a new FirestoreChannel object
<img width="867" alt="image" src="https://user-images.githubusercontent.com/17964075/167250520-72cdea81-d801-4946-b374-bf9718e8a271.png">
4. create a new GrpcCallProvider object
<img width="903" alt="image" src="https://user-images.githubusercontent.com/17964075/167250551-ca83b325-fbfa-4309-9cf8-764b25332215.png">
5. `GrpcCallProvider.initChannelTask()` -> `GrpcCallProvider#initChannel(Context context, DatabaseInfo databaseInfo)` -> ManagedChannelBuilder.forTarget(databaseInfo.getHost())
<img width="888" alt="image" src="https://user-images.githubusercontent.com/17964075/167250665-8ce890f0-4945-4dde-86d9-fabd82fc55d7.png">
6. `ManagedChannelProvider.provider().builderForTarget(target)` -> `ManagedChannelProvider.provider()`
<img width="704" alt="image" src="https://user-images.githubusercontent.com/17964075/167250722-9f8e602c-ba13-43b7-a9ae-ee79e5bdced9.png">
7. finally, the crash happen
<img width="877" alt="image" src="https://user-images.githubusercontent.com/17964075/167250745-c0085712-7a26-4aed-97cd-a729cf856f9e.png">

## question
what cause `ManagedChannelRegistry.getDefaultRegistry().provider()` get null?
Does it depends on the environment of the user's android mobile phone? such as 'there is no network?‘

how can i slove this problem ? can anyone help me?
I need help , thx

**Comment by sanjaypujare** [comment]
As the message suggests can you make sure you have grpc-okhttp as a dependency in your application?

**Comment by cxMax** [comment]
> As the message suggests can you make sure you have grpc-okhttp as a dependency in your application?

thank for ur reply.

## dependencies
below is the dependencies of firebase-firestore, I found grpc-okhttp is existed, 🤔️

![image](https://user-images.githubusercontent.com/17964075/167359976-5d7bbe19-fcfd-496c-abc4-ed5fe44f3700.png)

**Comment by sanjaypujare** [comment]
The dependency io.grpc:grpc-core:[1.41.0] has FAILED next to it. Do you have that one present?

**Comment by cxMax** [comment]
> The dependency io.grpc:grpc-core:[1.41.0] has FAILED next to it. Do you have that one present?
Thanks for pointing out what caused the build to fail

## I rebuild the dependencies of firebase-firestore
<img width="690" alt="image" src="https://user-images.githubusercontent.com/17964075/167572115-f188d071-dc0b-4fd8-9b6a-a2efa7668146.png">

## run ./gradlew dependencies :app:dependencies

![image](https://user-images.githubusercontent.com/17964075/167573159-214f595d-0d68-4c90-9c3e-4cbe9457480b.png)

![image](https://user-images.githubusercontent.com/17964075/167573080-070ac7e1-d6c2-4e7b-a910-981e982cab37.png)

## in conclusion
I don't know whether it works, 
I'll be back and update the final conclusion if it works,
Thanks again for your suggestion.
